iree/compiler/Conversion/LinalgToSPIRV/LinalgTileAndFusePass.cpp

```diff
@@ -166,15 +163,20 @@ LogicalResult TileSizeCalculator::inferTileAndWorkgroupSize(
     workgroupSize = {tileSizeX, tileSizeY, 1};
     return success();
   }
-  if (opInfo & OpInfo::Matmul) {
-    // TODO: For now just hard wire this, but we can do better.
-    tileSizes = {8, 8, 4};
-    workgroupSize = {8, 8, 1};
-    return success();
-  }
-  if (opInfo & OpInfo::BatchMatmul) {
-    tileSizes = {2, 8, 8, 4};
-    workgroupSize = {8, 8, 2};
+  if ((opInfo & OpInfo::Matmul) || (opInfo & OpInfo::BatchMatmul)) {
+    // Set tile size {M, N, K} = {maxWorkgroupSize /
+    // ceil(sqrt(maxWorkgroupSize)), ceil(sqrt(maxWorkgroupSize)), 32}
+    int64_t tileSizeN = std::max<int64_t>(
+        1, static_cast<int64_t>(llvm::PowerOf2Ceil(
+               static_cast<uint64_t>(std::sqrt(maxWorkgroupSize)))));
+    int64_t tileSizeM = maxWorkgroupSize / tileSizeN;
+    /// Hardwire tileSizeK, but would be determined from workgroup memory usage.
+    int64_t tileSizeK = 32;
+    if (opInfo & OpInfo::BatchMatmul)
+      tileSizes = {1, tileSizeM, tileSizeN, tileSizeK};
+    else
+      tileSizes = {tileSizeM, tileSizeN, tileSizeK};
+    workgroupSize = {tileSizeN, tileSizeM, 1};
```

Comment:
Is this really the final workgroup size? I'm not sure I follow the relationship between tile size, and workgroup size. I think they should be independent as we don't want to create large workgroups.

Reply:
I think the workgroup size is

1) taking tileSizes from beginning until hit a non-parallel loop. In this case, BatchMatmul is (1, M, N) and Matmul is {M, N}
2) reverse the 1)'s result, BatchMatmul is {N, M, 1} and Matmul is {M, N}/
3) Truncate to size 3 and pad with value `1`.

All the workgroup sizes are {N, M, 1}.